ling_flood.py

```diff
@@ -171,19 +164,39 @@ def __init__(self):
          random and have to check queen position constantly(which might be slow), probably
          will have to be changed too with the introduction of creepy spread queens"""
         queens = self.units(QUEEN)
-        hatchery = self.units(HATCHERY)
-        if hatchery.exists:
-            hatcheries_random = self.townhalls.random
-            if self.units(SPAWNINGPOOL).ready.exists:
-                if queens.amount < hatchery.amount and hatcheries_random.is_ready \
-                    and not self.already_pending(QUEEN)\
-                    and hatcheries_random.noqueue and self.supply_left > 1:
-                    if self.can_afford(QUEEN) and not queens.closer_than(8, hatcheries_random):
-                        self.actions.append(hatcheries_random.train(QUEEN))
-            for queen in queens.idle:
-                selected = self.townhalls.closest_to(queen.position)
-                if queen.energy >= 25 and not selected.has_buff(QUEENSPAWNLARVATIMER):
-                    self.actions.append(queen(EFFECT_INJECTLARVA, selected))
+        bases = self.units(HATCHERY) | self.units(LAIR) | self.units(HIVE)
+
+        # Handle injection
+        for base in bases:
+            if base.tag in self.base_queen_relation:
+                queen = self.base_queen_relation[base.tag]
+
+                if not queen: # If queen is dead
+                    del self.base_queen_relation[base.tag]
+
+                if queen.energy >= 25 and not base.has_buff(QUEENSPAWNLARVATIMER):
+                    self.actions.append(queen(EFFECT_INJECTLARVA, base))
+
+        queen_with_no_base = queens.filter(lambda queen: queen.tag not in map(lambda unit: unit.tag, self.base_queen_relation.values()))

```

Comment:
filter lambda is slow,  but with the way this library was built it can't be avoided, but map lambda can certainly be replaced:
`queen_with_no_base = queens.filter(lambda queen: queen.tag not in[unit.tag for unit in self.base_queen_relation.values()])`
Might work